// ngx-text-diff/esm5/lib/ngx-text-diff.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function DiffTableFormatOption() { }
if (false) {
    /** @type {?} */
    DiffTableFormatOption.prototype.id;
    /** @type {?} */
    DiffTableFormatOption.prototype.name;
    /** @type {?} */
    DiffTableFormatOption.prototype.label;
    /** @type {?} */
    DiffTableFormatOption.prototype.value;
    /** @type {?|undefined} */
    DiffTableFormatOption.prototype.icon;
    /** @type {?|undefined} */
    DiffTableFormatOption.prototype.disabled;
}
/**
 * @record
 */
export function DiffContent() { }
if (false) {
    /** @type {?} */
    DiffContent.prototype.leftContent;
    /** @type {?} */
    DiffContent.prototype.rightContent;
}
/**
 * @record
 */
export function DiffPart() { }
if (false) {
    /** @type {?} */
    DiffPart.prototype.content;
    /** @type {?} */
    DiffPart.prototype.isDiff;
}
/**
 * @record
 */
export function DiffLineResult() { }
if (false) {
    /** @type {?} */
    DiffLineResult.prototype.lineNumber;
    /** @type {?} */
    DiffLineResult.prototype.prefix;
    /** @type {?} */
    DiffLineResult.prototype.lineContent;
    /** @type {?} */
    DiffLineResult.prototype.lineDiffs;
}
/**
 * @record
 */
export function DiffTableRowResult() { }
if (false) {
    /** @type {?} */
    DiffTableRowResult.prototype.leftContent;
    /** @type {?} */
    DiffTableRowResult.prototype.rightContent;
    /** @type {?} */
    DiffTableRowResult.prototype.belongTo;
    /** @type {?} */
    DiffTableRowResult.prototype.hasDiffs;
    /** @type {?} */
    DiffTableRowResult.prototype.numDiffs;
}
/**
 * @record
 */
export function DiffResults() { }
if (false) {
    /** @type {?} */
    DiffResults.prototype.hasDiff;
    /** @type {?} */
    DiffResults.prototype.diffsCount;
    /** @type {?} */
    DiffResults.prototype.rowsWithDiff;
}
